### PR TITLE
Add an old machine cache to ArbCore to speed up reorgs

### PR DESCRIPTION
**Pull request by PlasmaPower**

### CODE REVIEW

File: packages/arb-avm-cpp/data_storage/src/arbcore.cpp [hunk 1]
```diff
@@ -611,6 +634,9 @@ template std::unique_ptr<MachineThread> ArbCore::getMachineUsingStateKeys(
     const MachineStateKeys& state_data,
     ValueCache& value_cache) const;
 
+constexpr uint256_t old_machine_cache_interval = 100'000;
+constexpr size_t old_machine_cache_max_size = 10;
```
Comment (hkalodner) [inline packages/arb-avm-cpp/data_storage/src/arbcore.cpp:638]: I'm concerned that a 1 million gas won't be enough of a reorg window. Maybe 1 million and 20 would be safer? There's a bit of a measurement question here. My concern is mainly about having to jump back to before the window and ending up loading a machine. I'm way less worried about doing some duplicate computation

**Review by hkalodner — CHANGES_REQUESTED** [review]

File: packages/arb-avm-cpp/data_storage/src/arbcore.cpp [hunk 2]
```diff
@@ -335,48 +335,71 @@ rocksdb::Status ArbCore::reorgToMessageCountOrBefore(
...
-                                << "Error loading machine from checkpoint: "
-                                << e.what() << std::endl;
-                            assert(false);
-                        }
+        // Delete each old cached machine until at or below message_sequence_number
+        {
+            std::unique_lock<std::shared_mutex> guard(old_machine_cache_mutex);
+            auto it = old_machine_cache.end();
+            while (it != old_machine_cache.begin()) {
+                it--;
```
[the remark this answers is not in the document]
Comment (PlasmaPower) [reply packages/arb-avm-cpp/data_storage/src/arbcore.cpp:343]: Does erase support reverse iterators? I was worried about that.

[the remark this answers is not in the document]
Comment (PlasmaPower) [reply packages/arb-avm-cpp/data_storage/src/arbcore.cpp:343]: It does not:
```
arbitrum/packages/arb-avm-cpp/data_storage/src/arbcore.cpp:355:44: error: no matching function for call to ‘std::map<intx::uint<256>, std::unique_ptr<Machine> >::erase(std::reverse_iterator<std::_Rb_tree_iterator<std::pair<const intx::uint<256>, std::unique_ptr<Machine> > > >&)’
  355 |             it = old_machine_cache.erase(it);
```

[the remark this answers is not in the document]
Comment (hkalodner) [reply packages/arb-avm-cpp/data_storage/src/arbcore.cpp:344]: Ah neat

**Review by hkalodner — APPROVED** [review]
LGTM

**Review by joshuacolvin0 — APPROVED** [review]
LGTM

**Pull request merged**